waku/waku_archive/driver/postgres_driver/postgres_driver.nim

```diff
@@ -587,6 +534,206 @@ proc sleep*(s: PostgresDriver, seconds: int):
+  if not partitionNamesRes.isOk():
+    return err("Could not retrieve partitions list: " & $partitionNamesRes.error)
+  else:
+    let partitionNames = partitionNamesRes.get()
+    for partitionName in partitionNames:
+      ## partitionName contains something like 'messages_1708449815_1708449875'
+      let bothTimes = partitionName.replace("messages_", "")
+      let times = bothTimes.split("_")
+      if times.len != 2:
+        return err(fmt"loopPartitionFactory wrong partition name {partitionName}")
+
+      var beginning: int64
+      try:
+        beginning = parseInt(times[0])
+      except ValueError:
+        return err("Could not parse beginning time: " & getCurrentExceptionMsg())
+
+      var `end`: int64
+      try:
+        `end` = parseInt(times[1])
+      except ValueError:
+        return err("Could not parse end time: " & getCurrentExceptionMsg())
+
+      self.partitionMngr.addPartitionInfo(partitionName, beginning, `end`)
+
+    return ok()
+
+const DefaultDatabasePartitionCheckTimeInterval = timer.minutes(10)
+const PartitionsRangeInterval = timer.hours(1) ## Time range covered by each parition
+
+proc loopPartitionFactory(self: PostgresDriver,
+                          onFatalError: OnFatalErrorHandler) {.async.} =
+  ## Loop proc that continuously checks whether we need to create a new partition.
+  ## Notice that the deletion of partitions is handled by the retention policy modules.
+
+  debug "starting loopPartitionFactory"
+
+  if PartitionsRangeInterval < DefaultDatabasePartitionCheckTimeInterval:
+    onFatalError("partition factory partition range interval should be bigger than check interval")
+
+  ## First of all, let's make the 'partition_manager' aware of the current partitions
+  (await self.initializePartitionsInfo()).isOkOr:
+    onFatalError("issue in loopPartitionFactory: " & $error)
+
+  while true:
+    trace "Check if we need to create a new partition"
+
+    let now = times.now().toTime().toUnix()
+
+    if self.partitionMngr.isEmpty():
+      debug "adding partition because now there aren't more partitions"
+      (await self.addPartition(now, PartitionsRangeInterval)).isOkOr:
+        onFatalError("error when creating a new partition from empty state: " & $error)
+    else:
+      let newestPartitionRes = self.partitionMngr.getNewestPartition()
+      if newestPartitionRes.isErr():
+        onFatalError("could not get newest partition: " & $newestPartitionRes.error)
+
+      let newestPartition = newestPartitionRes.get()
+      if newestPartition.containsMoment(now):
+        debug "creating a new partition for the future"
+        ## The current used partition is the last one that was created.
+        ## Thus, let's create another partition for the future.
+        (await self.addPartition(newestPartition.getLastMoment(), PartitionsRangeInterval)).isOkOr:
+          onFatalError("could not add the next partition for 'now': " & $error)
+      elif now >= newestPartition.getLastMoment():
```

Comment:
Really like this idea of creating always one partition ahead of time.

In that case, there might be partitions without messages at all (if we create it in advance and don't receive anything in the defined period).

If that happens we should probably remove it?

Reply:
> Do we expect this scenario regular? I mean is it possible that we would like to store message while we do not have a right time-range partition prepared for it yet?

Yes, this could happen for example if a node has been stopped for a while and gets connected again to the database (considering only one node connected to the database.)

----

> I feel more safe that rather create partition always one period ahead of become actual. WDYT?

Sorry, I can't quite understand this comment

Reply:
> If that happens we should probably remove it?

In this case, the retention policy will eat the older partitions until it gets up to the desired size. If one partition is empty, then it will need to truncate more partitions until reaching the target size.